fms/modules/tp.py

```diff
@@ -31,54 +25,229 @@ def setup_tp(self, rank: int, world_size: int) -> None:
+        param.copy_(tensor, non_blocking=True)
+
+    def _get_sd_weight(
+        self, state_dict: Dict[str, torch.Tensor], conditions: List[str]
+    ):
+        results = []
+        for k in state_dict:
+            all_conds = True
+            for cond in conditions:
+                if cond not in k:
+                    all_conds = False
+            if all_conds:
+                results.append(state_dict[k])
+        if len(results) == 1:
+            return results[0]
+        elif len(results) > 1:
+            raise ValueError(
+                f"Conditions not stringent enough: keys are {', '.join(state_dict.keys())}"
             )
-        tp_sharded_modules = list(
-            itertools.chain(
-                self.colwise_param_names(),
-                self.rowwise_param_names(),
-                self.embedding_param_names(),
+        else:
+            raise ValueError(
+                f"Weight not found, weights names are {', '.join(state_dict.keys())}"
             )
-        )
-        with torch.no_grad():
-            for mod_name, module in self.named_children():
-                if not mod_name in tp_sharded_modules:
-                    for param_name, param in module.named_parameters(recurse=False):
-                        param.copy_(
-                            getattr(getattr(module, mod_name), param_name),
-                            non_blocking=True,
-                        )
+
+    def load_weights(
+        self,
+        tensor_values: Dict[str, torch.Tensor],
+    ):
+        """Load all tensor values into a TP module.
+
+        Override this method to load weights into a TP module. You can see
+        examples for all TP modules in FMS, but the functions will generally
+        have the following structure:
+
+        1. Grab the necessary weights from tensor_values. Which weights
+            these are depend on each module.
+        2. Raise exceptions for missing required weights in tensor_values
+            (ValueError) or for unused weights in tensor_values
+            (AttributeError)
+        3. Use one of the copy_{rowwise,colwise} methods to load and shard
+            each weight correctly into the TP module. All the copy_ methods
+            explain in detail how to call them properly. Pay special attention
+            to the max_partition_sizes parameter for supporting fused weights
+            and MQA/GQA among others.
+
+        PSA: Each weight has a List[int] (max_partition_sizes) associated
+        with it where if the list is larger than size 1, this implies the
+        weight is fused (where the length of the list is the number of fused
+        weights in a single parameter).
+
+        Args:
+            tensor_values: Dict[str, torch.Tensor]
+                a state dict containing all the weights for a TP module
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def import_module(module, group: ProcessGroup):
+        pass
```

Comment:
is import_module still used or should we just remove it? cc @ani300

Reply:
yes, still used. Its function is to create the right sized TPModule from its non-TP counterpart + rank&world_size. Now that we centralized all the weight loading in `load_weights`, these are the only 2 functions that need overriding in each submodule: one to create the correct storage, one to fill it correctly.